contracts/Consensus.sol

```diff
@@ -0,0 +1,180 @@
+pragma solidity 0.4.24;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+contract Consensus {
+  using SafeMath for uint256;
+
+  event InitiateChange(bytes32 indexed parentHash, address[] newSet);
+  event ChangeFinalized(address[] newSet);
+
+  struct ValidatorState {
+    // Is this a validator.
+    bool isValidator;
+    // Is a validator finalized.
+    bool isValidatorFinalized;
+    // Indexes in the currentValidators.
+    uint256[] indexes;
+  }
+
+  address public SYSTEM_ADDRESS = 0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE;
+
+  bool public finalized = false;
+  address public owner;
+  uint256 public minStake;
+
+  address[] public currentValidators;
+  address[] public pendingValidators;
+  mapping(address => ValidatorState) public validatorsState;
+
+  mapping (address => uint256) public stakeAmount;
+
+  modifier onlySystem() {
+    require(msg.sender == SYSTEM_ADDRESS);
+    _;
+  }
+
+  modifier onlyOwner() {
+    require (msg.sender == owner);
+    _;
+  }
+
+  modifier notFinalized() {
+    require (!finalized);
+    _;
+  }
+
+  constructor(uint256 _minStake) public {
+    owner = msg.sender;
+    require (_minStake > 0);
+    setMinStake(_minStake);
+  }
+
+  function setMinStake(uint256 _minStake) public onlyOwner {
+    minStake = _minStake;
+  }
+
+  function getValidators() public view returns(address[]) {
+    return currentValidators;
+  }
+
+  function getPendingValidators() public view returns(address[]) {
+    return pendingValidators;
+  }
+
+  function finalizeChange() public onlySystem notFinalized {
+    finalized = true;
+
+    for (uint256 i = 0; i < pendingValidators.length; i++) {
+      ValidatorState storage state = validatorsState[pendingValidators[i]];
+      if (!state.isValidatorFinalized) {
+          state.isValidatorFinalized = true;
+      }
+    }
+
+    currentValidators = pendingValidators;
+
+    emit ChangeFinalized(getValidators());
+  }
+
+  function () external payable {
+    _stake(msg.sender, msg.value);
+  }
+
+  function withdraw(uint256 _amount) external {
+    require (_amount > 0);
+    require (_amount <= stakeAmount[msg.sender]);
+
+    stakeAmount[msg.sender] = stakeAmount[msg.sender].sub(_amount);
+
+    if(_amount >= minStake) {
```

Comment:
if minStake = 1000, and I staked 1001 coins I''ll get 1 validator. Then I withdraw 999 coins, and will still have one validator cause 999 < 1000. 

Then I'll do this process again 1000 times and I'll get 1000 validators, staking 2 coins in each one :sweat_smile:

Reply:
True :) will fix